File: ui/packages/sim-core/src/mds.ts
import { eigSymmetric, zeros, type Matrix } from "./linalg";

function symmetrize(D: Matrix): Matrix {
  const n = D.length;
  const out = zeros(n, n);
  for (let i = 0; i < n; i++) {
    const rowI = D[i]!;
    const rowOut = out[i]!;
    for (let j = 0; j < n; j++) {
      rowOut[j] = (rowI[j]! + D[j]![i]!) / 2;
    }
  }
  return out;
}

function doubleCenter(D: Matrix): Matrix {
  const n = D.length;
  const Dsym = symmetrize(D);
  const D2 = zeros(n, n);
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      const v = Dsym[i]![j]!;
      D2[i]![j] = v * v;
    }
  }
  const rowMean = new Array<number>(n).fill(0);
  const colMean = new Array<number>(n).fill(0);
  let totalMean = 0;
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      const v = D2[i]![j]!;
      rowMean[i]! += v;
      colMean[j]! += v;
      totalMean += v;
    }
  }
  for (let i = 0; i < n; i++) {
    rowMean[i]! /= n;
    colMean[i]! /= n;
  }
  totalMean /= n * n;
  const B = zeros(n, n);
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      B[i]![j] = -0.5 * (D2[i]![j]! - rowMean[i]! - colMean[j]! + totalMean);
    }
  }
  return B;
}

export function embeddabilityScore(D: Matrix): number {
  if (D.length < 3) return 0.0;
  const B = doubleCenter(D);
  const { values } = eigSymmetric(B);
  const sorted = values.slice().sort((a, b) => Math.abs(b) - Math.abs(a));
  const topTwo = Math.abs(sorted[0]!) + Math.abs(sorted[1]!) + 1e-12;
  let non2d = 0;
  for (let k = 2; k < sorted.length; k++) non2d += Math.abs(sorted[k]!);
  return non2d / topTwo;
}

export function perPointResiduals(D: Matrix): number[] {
  const n = D.length;
  if (n < 3) return new Array<number>(n).fill(0);
  const B = doubleCenter(D);
  const { values, vectors } = eigSymmetric(B);
  const order = values
    .map((v, i) => ({ v, i }))
    .sort((a, b) => Math.abs(b.v) - Math.abs(a.v));
  const residuals = new Array<number>(n).fill(0);
  for (let rank = 2; rank < n; rank++) {
    const idx = order[rank]!.i;
    const lambda = Math.abs(order[rank]!.v);
    for (let i = 0; i < n; i++) {
      const comp = vectors[i]![idx]!;
      residuals[i]! += comp * comp * lambda;
    }
  }
  return residuals;
}

export function embed2D(D: Matrix): {
  points: Array<[number, number]>;
  eigenvalues: [number, number];
  embeddability: number;
} {
  const n = D.length;
  if (n < 2) return { points: [], eigenvalues: [0, 0], embeddability: 0 };
  const B = doubleCenter(D);
  const { values, vectors } = eigSymmetric(B);
  const order = values
    .map((v, i) => ({ v, i }))
    .sort((a, b) => Math.abs(b.v) - Math.abs(a.v));
  const top1 = order[0]!;
  const top2 = order[1] ?? top1;
  const lambda1 = Math.max(0, top1.v);
  const lambda2 = Math.max(0, top2.v);
  const s1 = Math.sqrt(lambda1);
  const s2 = Math.sqrt(lambda2);
  const points: Array<[number, number]> = [];
  for (let i = 0; i < n; i++) {
    points.push([vectors[i]![top1.i]! * s1, vectors[i]![top2.i]! * s2]);
  }
  const sortedAbs = order.map((o) => Math.abs(o.v));
  const topTwoSum = sortedAbs[0]! + (sortedAbs[1] ?? 0) + 1e-12;
  let nonPlanar = 0;
  for (let k = 2; k < sortedAbs.length; k++) nonPlanar += sortedAbs[k]!;
  const embeddability = nonPlanar / topTwoSum;
  return { points, eigenvalues: [lambda1, lambda2], embeddability };
}

export function lyingEdgeResiduals(D: Matrix): Matrix {
  const n = D.length;
  if (n < 3) return zeros(n, n);
  const Dsym = symmetrize(D);
  const residuals = zeros(n, n);
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      if (i === j) continue;
      let acc = 0;
      for (let k = 0; k < n; k++) {
        if (k === i || k === j) continue;
        const excess = Dsym[i]![j]! - Dsym[i]![k]! - Dsym[j]![k]!;
        if (excess > 0) acc += excess;
      }
      residuals[i]![j] = acc;
    }
  }
  return residuals;
}

File: src/specter/trust/mds.py
"""Eigenvalue-residual classical MDS for Tier 2 cohort consistency. ADR 0015.

For a cohort of `k ≥ 3` peers, build the symmetric pairwise distance matrix `D`
from reciprocal beacon ranges. Classical MDS (Torgerson) recovers a Euclidean
embedding; if the matrix is consistent with a 2D placement, only the top two
eigenvalues of the doubly-centered Gram matrix `B = -½·J·D²·J` are large and
the rest are near zero. Non-2D-embeddability — caused by colluder-pair range
lies that pass pairwise reciprocal Tier 1 but disagree with the multilateration
geometry — surfaces as significant non-2D eigenvalue contribution.

Empirically validated (pre-PRD):

- N ∈ {4, 8, 12, 16, 20} honest swarms: `embeddability_score < 0.05`.
- Colluder pair (Tier-1-evading symmetric range lie): 9/9 trials at N up to 20
  correctly identified both colluders in the top-2 residual ranking. Ratio
  attack/honest score 1.5–8.3× across N.
- Compute: <500 μs per cohort at N=20 (sub-millisecond at the demonstrator
  swarm scale).

Public API:
- `embeddability_score(D)` — non-2D contribution / 2D contribution; 0 = perfect.
- `per_point_residuals(D)` — each peer's contribution to non-2D eigenvectors.
  The peer with the maximum residual is the outlier (in colluder-pair attacks,
  both colluders carry large residuals).

Internals symmetrize `D` before centering (`(D + D.T) / 2`) so asymmetric
range matrices — which Tier 1 would have already flagged — produce stable
MDS output for Tier 2's residual ranking.
"""

from __future__ import annotations

import numpy as np


def _double_center(D: np.ndarray) -> np.ndarray:
    """Compute the Gram matrix `B = -½·J·D²·J` (Torgerson centering).

    `J = I - (1/n)·ones(n,n)` removes the mean from each row/column of `D²`.
    For a Euclidean distance matrix, `B` has rank ≤ n−1 with the top
    eigenvalues equal to the coordinate variances along principal axes.
    """
    n = D.shape[0]
    D_sym = (D + D.T) / 2.0
    D_sq = D_sym ** 2
    J = np.eye(n) - np.ones((n, n)) / n
    B: np.ndarray = -0.5 * (J @ D_sq @ J)
    return B


def embeddability_score(D: np.ndarray) -> float:
    """Non-2D-embeddability score for the distance matrix `D`.

    Returns `sum(|λ_k|) for k ≥ 2  /  sum(|λ_0|, |λ_1|)`. A perfectly
    2D-embeddable matrix has score → 0. Larger values indicate the joint
    geometry cannot be realized in 2D — the Tier-2 collusion signal.

    Empirically, honest swarms produce scores < 0.05; colluder-pair attacks
    elevate to 0.05–0.5 depending on N (see ADR 0015).
    """
    if D.shape[0] < 3:
        return 0.0  # k=2 is trivially embeddable; Tier 2 needs k≥3
    B = _double_center(D)
    eigvals = np.linalg.eigvalsh(B)
    # eigvalsh returns ascending; we want descending by magnitude.
    eigvals = eigvals[np.argsort(-np.abs(eigvals))]
    top_two = np.abs(eigvals[:2]).sum() + 1e-12
    non_2d = np.abs(eigvals[2:]).sum()
    return float(non_2d / top_two)


def per_point_residuals(D: np.ndarray) -> np.ndarray:
    """Per-peer contribution to the non-2D eigenvectors.

    For each peer `i`, residual = `sum over k≥2 of (eigenvec_k[i])² · |λ_k|`.
    Surfaces "how movable" each peer is in the non-2D subspace. For *single*
    range-lying peers this aligns with the guilty peer; for *colluder-pair*
    attacks the bad edge is the load-bearing signal — see
    `lying_edge_residuals` for edge-level attribution.
    """
    n = D.shape[0]
    if n < 3:
        return np.zeros(n)
    B = _double_center(D)
    eigvals, eigvecs = np.linalg.eigh(B)
    idx = np.argsort(-np.abs(eigvals))
    eigvals = eigvals[idx]
    eigvecs = eigvecs[:, idx]
    residuals = np.zeros(n)
    for k in range(2, n):
        residuals += (eigvecs[:, k] ** 2) * abs(eigvals[k])
    return residuals


def lying_edge_residuals(D: np.ndarray) -> np.ndarray:
    """Per-edge triangle-inequality violation. For each edge `(i, j)`, sum
    triangle violations across all third peers `k`:

      `max(0, D[i,j] − D[i,k] − D[j,k])`

    A symmetric lie inflating `D[i,j]` by `bias` violates the triangle
    inequality against every `k` (because honest geometry has `D[i,j] ≤
    D[i,k] + D[j,k]`). Honest edges paired with a colluder show no violation
    on themselves — only the lying edge accumulates triangle violations.

    Empirically this localizes blame to the actual lying edge for
    Tier-1-evading symmetric colluder-pair attacks, where MDS-reconstruction
    residual disperses across edges.
    """
    n = D.shape[0]
    if n < 3:
        return np.zeros((n, n))
    D_sym = (D + D.T) / 2.0
    residuals = np.zeros((n, n))
    for i in range(n):
        for j in range(n):
            if i == j:
                continue
            for k in range(n):
                if k == i or k == j:
                    continue
                excess = D_sym[i, j] - D_sym[i, k] - D_sym[j, k]
                if excess > 0:
                    residuals[i, j] += excess
    return residuals
